### PR TITLE
created cards for each coin

### PR DESCRIPTION
**Pull request by Seif-Mamdouh**
Screenshot 2024-03-10 at 11 42 26 AM

### FetchCrypto.tsx 
controls the state of the API
has a async function handles error and return the API fetch promise in the useEffect
the component return all coins from an API

I also added a input tag for the user to search for any coin.
I need to add a Error boundary for just incase a coin is not found

### CryptoCard.tsx
has three props name, image and price for each Coin

### CODE REVIEW

File: src/app/components/FetchCryptoData.tsx [hunk 1]
```diff
@@ -1,29 +1,86 @@
...
+                          name={coin.name}
+                          image={coin.image}
+                          price={coin.current_price}
+                        />
+                      </div>
+                    );
+                  })}
             </div>
+          </>
+          // <div>
```
Comment (ming1in) [inline src/app/components/FetchCryptoData.tsx]: we should clean this up

File: src/app/components/SearchCrypto.tsx [hunk 2]
```diff
@@ -0,0 +1,5 @@
+// import React from 'react'
+
```
Comment (ming1in) [inline src/app/components/SearchCrypto.tsx]: do you still need this?

File: src/styles/crypto.css [hunk 3]
```diff
@@ -0,0 +1,15 @@
+.crypto-card {
+    border: 1px solid #ccc;
+    border-radius: 5px;
+    padding: 20px;
+    margin: 10px;
+    width: 200px; /* Adjust width as needed */
```
Comment (ming1in) [inline src/styles/crypto.css]: is this comment a TODO, that will be addressed in a follow up?

File: src/app/components/FetchCryptoData.tsx [hunk 4]
```diff
@@ -1,29 +1,86 @@
...
     const [state, setState] = React.useState({
         data: null,
         error: null
     });
 
+    const [search, setSearch] = React.useState("");
+
     React.useEffect(() => {
-        fetch("https://api.coingecko.com/api/v3/coins/markets?vs_currency=usd")
+        console.log(state.data)
```
Comment (ming1in) [inline src/app/components/FetchCryptoData.tsx]: should we clean up this log?

File: src/app/components/FetchCryptoData.tsx [hunk 5]
```diff
@@ -1,29 +1,86 @@
...
+        fetch(
+          "https://api.coingecko.com/api/v3/coins/markets?vs_currency=usd&order=market_cap_desc&per_page=100&page=1&sparkline=false"
```
Comment (ming1in) [inline src/app/components/FetchCryptoData.tsx]: we could pull this API fetch logic into a hook and DRY up this logic for other uses.

File: src/app/components/FetchCryptoData.tsx [hunk 6]
```diff
@@ -1,29 +1,86 @@
 import React from 'react'
 
+import CryptoCard from './CryptoCard';
+
```
Comment (ming1in) [inline src/app/components/FetchCryptoData.tsx]: nit: clean up extra newlines

File: src/styles/crypto.css [hunk 7]
```diff
@@ -0,0 +1,15 @@
+.crypto-card {
```
Comment (ming1in) [inline src/styles/crypto.css]: is this stylesheet being used?

File: src/app/components/CryptoCard.tsx [hunk 8]
```diff
@@ -0,0 +1,25 @@
+import React from 'react'
+
+
+
+
+const CryptoCard = ({image, name, price}) => {
```
Comment (ming1in) [inline src/app/components/CryptoCard.tsx]: can we get type definitions on the props here

File: src/app/components/FetchCryptoData.tsx [hunk 9]
```diff
@@ -1,29 +1,86 @@
...
+
+
 const FetchCryptoData = () => {
     const [state, setState] = React.useState({
```
Comment (ming1in) [inline src/app/components/FetchCryptoData.tsx]: we should add type definition to state objects

File: src/app/components/FetchCryptoData.tsx [hunk 10]
```diff
@@ -1,29 +1,86 @@
...
             (data) => setState({ ...state, data: data }),
             (error) => setState({ ...state, error: error })
           );
     }, []);
 
+    const handleChange = (e) => {
+        setSearch(e.target.value);
+    }
+
+      const filteredCoins = state.data && state.data.filter((coin) =>
```
Comment (ming1in) [inline src/app/components/FetchCryptoData.tsx]: nit: inconsistent tabs

File: src/app/components/FetchCryptoData.tsx [hunk 11]
```diff
@@ -1,29 +1,86 @@
...
+        coin.name.toLowerCase().includes(search.toLowerCase())
+      );
+
     if (state.error) {
-        return <div>Error: {state.error.message}</div>;
+        return <div>Error: {state.error.message} </div>;
     } else {
```
Comment (ming1in) [inline src/app/components/FetchCryptoData.tsx]: you dont need this `else{}` here

**Review by ming1in — CHANGES_REQUESTED** [review]
I dont have the full context here, but left some comments on code style and TS best practices

**Comment by ming1in** [comment]
your branch name should follow some industry convention

you could use this 
- https://dev.to/varbsan/a-simplified-convention-for-naming-branches-and-commits-in-git-il4

or lead with your name

`<your username>/<feature >`

File: src/styles/crypto.css (hunk 7 above)
Comment (Seif-Mamdouh) [reply src/styles/crypto.css]: nope

File: src/app/components/CryptoCard.tsx [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+  name: string,
+  price: number,
+}
+
+
+const CryptoCard = ( props: CryptoCardProps) => {
+    return (
+      <div
+        style={{
+                border: "1px solid #ccc",
```
Comment (ming1in) [inline src/app/components/CryptoCard.tsx:16]: in the future we should figure out how to pull these into a stylesheet

File: src/app/components/FetchCryptoData.tsx [hunk 13]
```diff
@@ -1,31 +1,53 @@
...
             <div>
-                {state.data && state.data.map((item: any) => (
-                    <div key={item.id}>{item.name}</div>
-                ))}
+                <input type="text" onChange={handleChange} />
             </div>
-        );
-    }
+            <div>
+                {filteredCoins && filteredCoins.map((coin) => {
```
Comment (ming1in) [inline src/app/components/FetchCryptoData.tsx]: I dont think you need this check `filteredCoins &&`. if `filteredCoins` is empty then nothing will render here

File: src/app/components/FetchCryptoData.tsx [hunk 14]
```diff
@@ -1,31 +1,53 @@
...
-    if (state.error) {
-        return <div>Error: {state.error.message}</div>;
-    } else {
-        return (
+    
+    const { data: state, error: error } = UseCoinGeckoApi();
+
+    const [search, setSearch] = React.useState("");
+
+    const handleChange = (e) => {
```
Comment (ming1in) [inline src/app/components/FetchCryptoData.tsx]: add a type definitions for `e` here

**Review by ming1in — COMMENTED** [review]
starting to looks much cleaner!!! 🔥

File: src/app/components/CryptoCard.tsx (hunk 12 above)
Comment (ming1in) [reply src/app/components/CryptoCard.tsx:16]: get into a habit of adding `TODO - ....` comments for tech debt

**Review by ming1in — APPROVED** [review]

**Pull request merged**